feat(regional): Italian Localization and E-Invoicing export

gaurav-naik opened the pull request:
**[WIP]**

![e-invoicing](https://user-images.githubusercontent.com/17795124/52044193-8d0fe880-2567-11e9-88e0-d17b720bdb28.gif)

### Setup
#### Company
1. Set the Tax ID and Fiscal Regime
2. Add an address (mandatory):
    1. Ensure that Address Line 1, City, State and Pincode fields have values
    2. The **State** field must have the standard province abbreviation. (RM for Rome, BR for Bari)

#### Customer
1. Ensure that Tax ID and Fiscal Code are correctly filled in.
    1. For Individuals, set a valid Fiscal Code.
    2. For Companies, set a valid Tax ID.
    3. For Individual Companies, set both Fiscal Code and Tax ID fields.
2. Add an address (mandatory):
    - Ensure that Address Line 1, City, State and Pincode fields have values
    - The State field must have the standard province abbreviation. (RM for Rome, BR for Bari)

#### Chart of Accounts 
1. Add accounts for VAT 0%, VAT 4%, VAT 10% as applicable.
2. For the VAT 0% account, set the Tax Exemption Reason (Natura) applicable to your company.

#### Mode of Payment
1. Select the appropriate code on the Mode of Payment doctype (MP02 for Cheque, for example)

#### Sales Invoice
1. Ensure that the TaxID and Addresses of the customer appear correctly on the invoice.
2. Set the Mode of Payment in the Payment Schedule.

### E-Invoice export:
1. Open report Electronic Invoicing Register via the Awesome Bar.
2. Set the desired filters.
3. Click Export E-Invoices to get a .zip file containing XML files for each submitted invoice.
4. Upload this .zip file to SDI or to a verified service provider.

https://github.com/frappe/erpnext/issues/16259

Commit:
feature(regional): Italian Localization

Added setup.py for italy under regional
Included fiscal regimes, tax exemption reasons and custom fields

Issue #16259

Commit:
mod(regional,italy): Updates to Italian Localization

fixes:
removed cleanup_files from italy/utils
removed extract_doc_number from italy/utils
added country filter to italian localization patch
replaced dict assignment with dot operator in prepare_invoice
correcions in e-invoice xml following changes in prepare_invoice
added setup_report in patch
added section for e-invoicing custom fields in company
added Italy in test_company

rohitwaghchaure commented on erpnext/regional/italy/utils.py:
```diff
@@ -0,0 +1,279 @@
+
+def cleanup_files(files):
+	#TODO: Clean up XML files after ZIP gets downloaded
+	pass
```
What is the use of this method?

gaurav-naik replied:
It is not relevant anymore. It can be rewritten as a function to cleanup zip files exported via Electronic Invoice Register as a daily task.

rohitwaghchaure commented on erpnext/regional/italy/utils.py:
```diff
@@ -0,0 +1,279 @@
+		else:
+			return name_parts[-1:][0]
+	else:
+		return doc.name
```
Also extract_doc_number method has not called from anywhere, is it required?

gaurav-naik replied:
Will drop.

rohitwaghchaure commented on erpnext/patches/v11_0/make_italian_localization_fields.py:
```diff
@@ -0,0 +1,8 @@
+from erpnext.regional.italy.setup import make_custom_fields
+
+def execute():
+    make_custom_fields()
```
This patch should only run for the Italy country

rohitwaghchaure commented on erpnext/regional/italy/utils.py:
```diff
@@ -0,0 +1,279 @@
+		if item.tax_rate == 0.0:
+			item["tax_exemption_reason"] = tax_data["0.0"]["tax_exemption_reason"]
+
+	invoice["invoice_items"] = items
```
In template you can use doc.items? If yes then no need of it, also for the payment_terms

Commit:
mod(regional,italy): adjusted calculation logic for charges

validation for missing code on mode of payment
UOM from settings
Replace str with cstr

rohitwaghchaure commented on erpnext/regional/italy/utils.py:
```diff
@@ -0,0 +1,285 @@
+from __future__ import unicode_literals
+
+import frappe, json, os
+from frappe.utils import flt
```
Add cstr and use cstr instead of str
